### PR TITLE
Update Client Add and Find Function

### LINKED ISSUE #135: Update cadd to allow same name. Change find duplicate person functionality
Allow same name.
Duplicate person only via 3 conditions:

1. Same name + same email
2. Same name + same phone number
3. Same name + same email + same phone number

### LINKED ISSUE #136: Update Find Command to find rental information

### LINKED ISSUE #148: When find nobody, MUST clear the rental information pane also

### LINKED ISSUE #149: Find command to allow finding substring

### LINKED ISSUE #150: Add tests wrt finding rental information

### PR DESCRIPTION
**Pull request by winstonlimjy**
Changes:
- Allow client to have same name. Fixes #135 
- Check for duplicate with all 3 fields: name + phone + email
- Finding of rental information, all fields. Fixes #136 
- Allow finding of substrings. Fixes #149 
- Create a new contains method in StringUtil.java 
- Create test cases for finding rental information. Fixes #150 
- Display no rental information when find function found nothing. Fixes #148

### CODE REVIEW

File: src/main/java/seedu/address/model/client/RentalInformationContainsKeywordsPredicate.java [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+import java.util.function.Predicate;
+
+import seedu.address.commons.util.StringUtil;
+import seedu.address.commons.util.ToStringBuilder;
+
+/**
+ * Tests that a {@code Client}'s {@code Rental Information} matches any of the keywords given.
+ */
+public class RentalInformationContainsKeywordsPredicate implements Predicate<Client> {
+
```
Comment (gohsl99) [inline src/main/java/seedu/address/model/client/RentalInformationContainsKeywordsPredicate.java:13]: Should remove redundant newline if possible to maintain consistency. Same for line 27 and 31.

File: src/test/java/seedu/address/logic/commands/FindCommandTest.java [hunk 2]
```diff
@@ -68,24 +79,116 @@ public void execute_zeroKeywords_noPersonFound() {
...
+        FindCommand command = new FindCommand(namePredicate, phonePredicate, emailPredicate, rentalInfoPredicate);
+        expectedModelWithRental.updateFilteredPersonList(
+                namePredicate.or(phonePredicate).or(emailPredicate).or(rentalInfoPredicate));
+        assertCommandSuccess(command, modelWithRental, expectedMessage, expectedModelWithRental);
+        assertEquals(expectedModelWithRental.getFilteredPersonList(), modelWithRental.getFilteredPersonList());
+    }
+
+    @Test
+    public void execute_monthlyRentKeyword_success() {
+        String address = "2900";
```
Comment (gohsl99) [inline src/test/java/seedu/address/logic/commands/FindCommandTest.java:110]: monthlyRent can be a better variable name here.

File: src/test/java/seedu/address/logic/commands/FindCommandTest.java [hunk 3]
```diff
@@ -68,24 +79,116 @@ public void execute_zeroKeywords_noPersonFound() {
...
+        FindCommand command = new FindCommand(namePredicate, phonePredicate, emailPredicate, rentalInfoPredicate);
+        expectedModelWithRental.updateFilteredPersonList(
+                namePredicate.or(phonePredicate).or(emailPredicate).or(rentalInfoPredicate));
+        assertCommandSuccess(command, modelWithRental, expectedMessage, expectedModelWithRental);
+        assertEquals(expectedModelWithRental.getFilteredPersonList(), modelWithRental.getFilteredPersonList());
+    }
+
+    @Test
+    public void execute_depositKeyword_success() {
+        String address = "5800";
```
Comment (gohsl99) [inline src/test/java/seedu/address/logic/commands/FindCommandTest.java:127]: deposit can be a better variable name here.

File: src/test/java/seedu/address/logic/commands/FindCommandTest.java [hunk 4]
```diff
@@ -68,24 +79,116 @@ public void execute_zeroKeywords_noPersonFound() {
...
+        FindCommand command = new FindCommand(namePredicate, phonePredicate, emailPredicate, rentalInfoPredicate);
+        expectedModelWithRental.updateFilteredPersonList(
+                namePredicate.or(phonePredicate).or(emailPredicate).or(rentalInfoPredicate));
+        assertCommandSuccess(command, modelWithRental, expectedMessage, expectedModelWithRental);
+        assertEquals(expectedModelWithRental.getFilteredPersonList(), modelWithRental.getFilteredPersonList());
+    }
+
+    @Test
+    public void execute_tenantsKeyword_success() {
+        String address = "carl";
```
Comment (gohsl99) [inline src/test/java/seedu/address/logic/commands/FindCommandTest.java:144]: tenant or customer can be a better variable name here.

File: src/test/java/seedu/address/logic/commands/FindCommandTest.java [hunk 5]
```diff
@@ -68,24 +79,116 @@ public void execute_zeroKeywords_noPersonFound() {
...
-    //        String expectedMessage = String.format(MESSAGE_PERSONS_LISTED_OVERVIEW, 3);
-    //        NameContainsKeywordsPredicate predicate = preparePredicate("Kurz Elle Kunz");
-    //        FindCommand command = new FindCommand(predicate);
-    //        expectedModel.updateFilteredPersonList(predicate);
-    //        assertCommandSuccess(command, model, expectedMessage, expectedModel);
-    //        assertEquals(Arrays.asList(CARL, ELLE, FIONA), model.getFilteredPersonList());
-    //    }
+    @Test
+    public void execute_rentalStartDateKeyword_success() {
+        String address = "01 Jan";
```
Comment (gohsl99) [inline src/test/java/seedu/address/logic/commands/FindCommandTest.java:161]: Same as above, should consider use rentalStartDate as variable name.

File: src/test/java/seedu/address/logic/commands/FindCommandTest.java [hunk 6]
```diff
@@ -68,24 +79,116 @@ public void execute_zeroKeywords_noPersonFound() {
...
+        FindCommand command = new FindCommand(namePredicate, phonePredicate, emailPredicate, rentalInfoPredicate);
+        expectedModelWithRental.updateFilteredPersonList(
+                namePredicate.or(phonePredicate).or(emailPredicate).or(rentalInfoPredicate));
+        assertCommandSuccess(command, modelWithRental, expectedMessage, expectedModelWithRental);
+        assertEquals(expectedModelWithRental.getFilteredPersonList(), modelWithRental.getFilteredPersonList());
+    }
+
+    @Test
+    public void execute_rentalEndDateKeyword_success() {
+        String address = "31 Dec";
```
Comment (gohsl99) [inline src/test/java/seedu/address/logic/commands/FindCommandTest.java:178]: Same as above, should consider use rentalEndDate as variable name.

File: src/test/java/seedu/address/logic/commands/FindCommandTest.java [hunk 7]
```diff
@@ -26,6 +28,8 @@ public class FindCommandTest {
     private Model model = new ModelManager(getTypicalAddressBook(), new UserPrefs());
     private Model expectedModel = new ModelManager(getTypicalAddressBook(), new UserPrefs());
 
+    private Model modelWithRental = new ModelManager(getTypicalAddressBookWithRental(), new UserPrefs());
+    private Model expectedModelWithRental = new ModelManager(getTypicalAddressBookWithRental(), new UserPrefs());
     @Test
     public void equals() {
         NameContainsKeywordsPredicate namePredicate1 =
```
Comment (gohsl99) [inline src/test/java/seedu/address/logic/commands/FindCommandTest.java:35]: May be good to change the name to something like firstNamePredicate to match with findFirstCommand used later.

**Review by gohsl99 — COMMENTED** [review]
Overall LGTM, minor code style can be changed.

File: src/main/java/seedu/address/model/client/RentalInformationContainsKeywordsPredicate.java (hunk 1 above)
Comment (winstonlimjy) [reply src/main/java/seedu/address/model/client/RentalInformationContainsKeywordsPredicate.java:13]: Thanks for pointing that out!

File: src/test/java/seedu/address/logic/commands/FindCommandTest.java (hunk 2 above)
Comment (winstonlimjy) [reply src/test/java/seedu/address/logic/commands/FindCommandTest.java:110]: Thanks for pointing that out!

File: src/test/java/seedu/address/logic/commands/FindCommandTest.java (hunk 3 above)
Comment (winstonlimjy) [reply src/test/java/seedu/address/logic/commands/FindCommandTest.java:127]: Thanks for pointing that out!

File: src/test/java/seedu/address/logic/commands/FindCommandTest.java (hunk 5 above)
Comment (winstonlimjy) [reply src/test/java/seedu/address/logic/commands/FindCommandTest.java:161]: Thanks for pointing that out!

File: src/test/java/seedu/address/logic/commands/FindCommandTest.java (hunk 6 above)
Comment (winstonlimjy) [reply src/test/java/seedu/address/logic/commands/FindCommandTest.java:178]: Thanks for pointing that out!

File: src/test/java/seedu/address/logic/commands/FindCommandTest.java (hunk 7 above)
Comment (winstonlimjy) [reply src/test/java/seedu/address/logic/commands/FindCommandTest.java:35]: I will keep the current name for now! I will keep in view for future changes! 

Thanks for pointing that out!

**Pull request merged**